substrate/client/rpc/src/state/state_full.rs

```diff
@@ -217,6 +237,20 @@ where
 			.map_err(client_err)
 	}
 
+	fn storage_diff(
+		&self,
+		block : Block::Hash,
+		prefixes: Option<Vec<StorageKey>>
+	) -> std::result::Result<(StorageCollection, ChildStorageCollection), Error> {
+		let (mut storage_collection, mut child_storage_collection) = self.client.storage_updates_at(block).map_err(client_err)?;
+
+		// retain only required prefixes
+		storage_collection.retain(|key| self.is_target_key(sc_client_api::StorageKey(key.0.clone()), prefixes.clone()));
+		child_storage_collection.retain(|key| self.is_target_key(sc_client_api::StorageKey(key.0.clone()), prefixes.clone()));
+
+		Ok((storage_collection, child_storage_collection))
+	}
+
```

Comment:
Since we have to build full keys anyway - let's return a single list of all changed storage elements.
Without splitting to storage_collection / child_storage_collection
I.e. `std::result::Result<Vec<(StorageKey, Option<StorageData>)>, JsonRpseeError>`
This is how the other RPC endpoint `storage_pairs` operates.